### DISCUSSION TITLE
Proposal: Avoid non-tail recursion

### DISCUSSION THREAD

**Discussion by Will-Low** [rust-unofficial/patterns#295] [Ideas]
Given that the Rust compiler does not current implement tail-call optimization, I think it would make sense to include non-tail recursion as an antipattern. Would folks agree?

Would it also make sense that, instead of rolling one's own tail-recursive functions, that we recommend the use of the `fold` method, which uses tail recursion?

**Comment by pickfire** [comment]
There are library that allows using tail recursion with macro, I forgot which one.
